Add order shipping rate endpoint

joshbetz opened the pull request:
This adds the initial boilerplate for the new shipping rates endpoint.
We're adding the route to the existing orders controller. The
permissions check is the same check that happens for a single order. The
only difference here is that we're adding a response handler
specifically for the shipping methods. It's not implemented right now,
but we need to calculate the cost for all available shipping methods for
the current order.

Closes #30794

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

We're adding a `shipping_methods` endpoint to the order object. This calculates the totals for each available shipping method, which can be useful for building a checkout flow. This allows us to tell customers what each method will cost so that they can make an informed decision about which one they want.

### How to test the changes in this Pull Request:

1. Create an order
2. GET /wp-json/wc/v3/orders/:id/shipping_methods
3. Observe that the shipping methods were correctly calculated

The response should look something like:

```
{
    "flat_rate:2": {
        "label": "Flat rate",
        "cost": 10
    },
    "local_pickup:3": {
        "label": "Local pickup",
        "cost": 1
    }
}
```

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

joshbetz committed:
Add order shipping rate endpoint

This adds the initial boilerplate for the new shipping rates endpoint.
We're adding the route to the existing orders controller. The
permissions check is the same check that happens for a single order. The
only difference here is that we're adding a response handler
specifically for the shipping methods. It's not implemented right now,
but we need to calculate the cost for all available shipping methods for
the current order.

Closes #30794

joshbetz commented:
One question I have here is that `WC_Shipping` seems to be tied pretty closely to `WC_Cart`. Do we need to instantiate a cart in order to calculate shipping rates?

https://github.com/woocommerce/woocommerce/blob/trunk/includes/class-wc-shipping.php#L245-L251

joshbetz committed:
Calculate shipping costs

Here we're adding all products to a cart and then using
`WC_Cart::calculate_shipping()` to calculate the cost of all shipping
methods.

The response format is similar to the `v3/shipping_methods` endpoint,
but adds the `cost` (since that's the point of this endpoint :D).

vedanshujain reviewed (commented):
I think a better approach may be possible by using `WC()->shipping()` like so:

1. We can use `WC()->shipping()->calculate_shipping()` to calculate available shipping methods and their costs. Trick is to build the `$packages` param correctly.
2. A good way to generate what the [package should expect is available here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-cart.php#L1494). We should also provide our own package key to not overwrite any session values.

joshbetz committed:
Change to WC()->shipping()->calculate_shipping() for calculating shipping

Per @vedanshujain's [feedback](https://github.com/woocommerce/woocommerce/pull/30855#pullrequestreview-789313637)
we're calculating the packages using the "cart" and then using
WC()->shipping() to calculate shipping.

Note: This still uses the in-memory session instead of the cookie. We're
loading up an order that is stored in the database and calculating
shipping for that.

joshbetz committed:
Calculate total shipping rate for each method

For each shipping method, we add the cost for each package to come up
with a total cost for each method.

Example response:

```
{
    "flat_rate:2": {
        "label": "Flat rate",
        "cost": 10
    },
    "local_pickup:3": {
        "label": "Local pickup",
        "cost": 1
    }
}
```

joshbetz committed:
Change the endpoint to shipping_methods for consistency

vedanshujain commented on plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-orders-v2-controller.php:
```diff
@@ -117,6 +117,27 @@ public function register_routes() {
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_shipping_methods' ),
+					'permission_callback' => array( $this, 'get_item_permissions_check' ),
```
This should also check for `wc_rest_check_manager_permissions( 'shipping_methods', 'read' ) ` and not just orders

joshbetz replied:
https://github.com/woocommerce/woocommerce/pull/30855/commits/ab49afbd5cb25dfa70a7e0faa7539551a5ef8b9d

vedanshujain commented on plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-orders-v2-controller.php:
```diff
@@ -132,6 +153,52 @@ public function register_routes() {
+		}
+
+		$cart = new WC_Cart();
+		foreach ( $order->get_items( 'line_item' ) as $item ) {
```
We should empty the cart before adding order items to prevent an already non-empty cart interfering with shipping calculations

joshbetz replied:
The cart only exists in memory, so for the API it will be destroyed after each request. Granted, unexpected things could happen if we use this code in WP-CLI, but that seems unlikely for a REST API endpoint?

https://github.com/woocommerce/woocommerce/pull/30855/files#diff-0546b54e79a59d3598b80bb065d59cd9dcc3ecb96f7618aba35c33805043cfa0R34

vedanshujain commented on plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-orders-v2-controller.php:
```diff
@@ -132,6 +153,52 @@ public function register_routes() {
+			$cart->add_to_cart( $item->get_product_id(), $item->get_quantity(), $item->get_variation_id() );
+		}
+
+		$packages = $cart->get_shipping_packages();
```
Note that this may potentially change the cart items. If this is supposed to be used by admin users then it should be fine, but we should empty the cart after our calculations are done.

Some other alternatives to not have this method interfere with cart items would be:
1. Writing a corresponding `get_shipping_packages` in `WC_Order` class as well. Maybe we can extract out this method in a trait and then use the same code in both places.

2. We can also look into saving cart contents in a temporary variable before adding our order items, and then restore it back once our calculations are done.

Both of these alternatives are high effort, so please take a call on whether we would want to do these or not based on the use case.

joshbetz replied:
In this case the cart only exists for the current request and will be reconstructed from the order every time, so I think it should be OK? In general, I think 1 is the best long term solution. The REST API should be stateless and it's somewhat unexpected that we're using a cart in this way.

vedanshujain reviewed (commented):
The general approach looks good to me, left some small comments. Seems like unit tests failures are unrelated?

joshbetz committed:
lint

Older versions of PHP don't support arrow functions

joshbetz committed:
Permissions: Check shipping_methods and orders

To check permissions on a shipping method, we need to ensure you have
access to both shiping methods and orders.

https://github.com/woocommerce/woocommerce/pull/30855#discussion_r806468245

joshbetz committed:
Fix permission check function name

joshbetz commented:
@samueljseay I would check with @coreymckrill about whether this is still needed.

samueljseay commented:
Thank you @joshbetz ! I'll find out.

coreymckrill commented:
I think this is still useful functionality, but I wonder if this is still the best approach given that there is now a Store API...

coreymckrill commented:
Ah, just went back and re-read the original issue, which already mentions the Store API. So I'd say this is still worth keeping.

barryhughes commented:
Closing, since this work has stalled and we're currently focused on other areas.
